capsules/src/spi_peripheral.rs

```diff
@@ -227,28 +252,44 @@ impl<S: SpiSlaveDevice> SpiSlaveClient for SpiPeripheral<'_, S> {
         length: usize,
     ) {
         self.app.map(move |app| {
-            if app.app_read.is_some() {
-                let src = readbuf.as_ref().unwrap();
-                let dest = app.app_read.as_mut().unwrap();
-                let start = app.index - length;
-                let end = start + length;
+            let rbuf = readbuf.map(|src| {
+                let index = app.index;
+                app.app_read.mut_map_or((), |dest| {
```

Comment:
🎏 , `mut_map_or` with unused return. Here is probably one of the more compelling cases for a plain `mut_map` to me: it's on the hardware event upcall path, but the app has disappeared. Not really meaningful to notify an app that's gone away about the spi event — maybe there's a case for trying to tear down the rest of the app's SPI configuration in response to this, but that feels like a black hole of complexity, especially since other, future, users are most likely to fixup the SPI bus anyway (so long as they can get access to it).

Reply:
What if the app hasn't disappeared, but rather it passed a new AppSlice that's shorter than the request length?

Line 292 gives me the willies -- it doesn't actually report how long the operation was.  Combined with the fact that lines 287 is an equality makes me think things could go very wrong. 

I'm planning on focusing on Tock tomorrow, I'll try to work through the cases here.

Reply:
I think the equality test on 287 is alright; the index is only set in `do_next_read_write`, and so it indicates "is this the final operation". However, if a slice is changed under the capsule, it's possible it could wedge. In particular, it's possible that

https://github.com/tock/tock/blob/a2317e5b2bc7311345e82d901db601cfe0733d25/capsules/src/spi_peripheral.rs#L67

resolves to 0, at which point the underlying SPI operation is for zero bytes. This could cause an error to be returned on 

https://github.com/tock/tock/blob/a2317e5b2bc7311345e82d901db601cfe0733d25/capsules/src/spi_peripheral.rs#L82

which is ignored. The capsule will sit waiting for a callback that never comes. 

For the particular code above, though, the issue is that you need line 

https://github.com/tock/tock/blob/a2317e5b2bc7311345e82d901db601cfe0733d25/capsules/src/spi_peripheral.rs#L257

to resolve to the actual number of total bytes written, so that it can be passed to a scheduled upcall. Otherwise, you have this weird situation of having 3 values:

1) The length passed to the command as arg1;
2) The length of the initial slice;
3) The length of the new slice.

Requiring the user program to keep track of all of these and which one will be used is asking too much, especially since the information is right here. So I'd argue that this should pass the value of `end` out of the closure.

The key thing about `map_or` isn't that it forces you to consider what if the process is dead: that's easy, do nothing with the buffer. It's that given these slices can change, you always need to pass back some information on what was actually done based on the current state of the the AppSlice. The problem with a standard `map` is that it leads to just ignoring the result (who wants to deal with an Option?).